Jimmba/aws-cdk
# Task10 bff

>link to the repo

https://github.com/Jimmba/aws-cdk

>Product Service API URL

https://hy98va2hj9.execute-api.eu-west-1.amazonaws.com/prod/products

>Cart Service API URL

https://q2pdox4j25.execute-api.eu-west-1.amazonaws.com/api/profile/cart

>BFF Service API URL

http://jimmba-bff-api-dev.eu-west-1.elasticbeanstalk.com/

>example of how to call createProduct lambda with all needed information: URL, payload, headers, etc.

url: http://jimmba-bff-api-dev.eu-west-1.elasticbeanstalk.com/product
method 'POST'
headers: `Authorization Basic {TOKEN}`
Token is encoded to base64 `{USER}:TEST_PASSWORD`, where USER - your github name

body example: 
```
{
    "description": "Description",
    "price": 999999999,
    "title": "Product",
    "count": 999999
}
```


> example how to call Product Service and Cart Service via BFF Service URL

http://jimmba-bff-api-dev.eu-west-1.elasticbeanstalk.com/product
http://jimmba-bff-api-dev.eu-west-1.elasticbeanstalk.com/cart (authorization header is required)

>A working and correct application should be in the bff-service folder. Reviewer can start this application locally with any valid configuration in the .env file and this application should works as described in the Task 10.1

To run project locally add .env file with variables `cart` and `product` with url of their services, for example:
```
product = 'https://hy98va2hj9.execute-api.eu-west-1.amazonaws.com/prod/products'
cart = 'https://q2pdox4j25.execute-api.eu-west-1.amazonaws.com/api/profile/cart'
```

You'll be able to call these endpoints:
http://localhost:4000/product
http://localhost:4000/cart

>The BFF Service should be deployed with Elastic Beanstalk. The BFF Service call should be redirected to the appropriate service : Product Service or Cart Service. The response from the BFF Service should be the same as if Product Service or Cart Service services were called directly.

![image](https://github.com/Jimmba/aws-cdk/assets/32487200/ea3c434b-d0fe-415f-afce-ab1aa164c311)

Additional (optional) tasks

+20 - Add a cache at the BFF Service level for a request to the getProductsList lambda function of the Product Service. The cache should expire in 2 minutes.t

How to test:
Get products list (METHOD 'GET' http://localhost:4000/product)
Create new product (METHOD 'POST' http://localhost:4000/product, other properties described above)
![image](https://github.com/Jimmba/aws-cdk/assets/32487200/f9c9e4fb-46ee-4317-9396-967ad54c35cd)

Get products list - result shouldn’t have new product
![image](https://github.com/Jimmba/aws-cdk/assets/32487200/70607164-acfd-44a4-b446-5bc457e6d080)

Wait more than 2 minutes
Get products list - result should have new product
![image](https://github.com/Jimmba/aws-cdk/assets/32487200/962b8ff9-dbad-4529-8ba4-02b14ba0ed7c)


Total: 100/100
